File: shared/save.ts
import { PetState, Screen, Stage, FLAG_SLEEPING } from "./types";
import { createNewPet } from "./pet";

export const SAVE_MAGIC = 0x54;
export const SAVE_VERSION = 1;
export const SAVE_SIZE = 17;

function crc8(data: Uint8Array, len: number): number {
  let crc = 0;
  for (let i = 0; i < len; i++) {
    crc ^= data[i];
    for (let b = 0; b < 8; b++) {
      crc = crc & 0x80 ? ((crc << 1) ^ 0x07) & 0xff : (crc << 1) & 0xff;
    }
  }
  return crc & 0xff;
}

function writeU32LE(buf: Uint8Array, offset: number, value: number): void {
  buf[offset] = value & 0xff;
  buf[offset + 1] = (value >>> 8) & 0xff;
  buf[offset + 2] = (value >>> 16) & 0xff;
  buf[offset + 3] = (value >>> 24) & 0xff;
}

function readU32LE(buf: Uint8Array, offset: number): number {
  return (
    buf[offset] |
    (buf[offset + 1] << 8) |
    (buf[offset + 2] << 16) |
    (buf[offset + 3] << 24)
  ) >>> 0;
}

/** Serialize to EEPROM-shaped blob (17 bytes). */
export function serialize(pet: PetState): Uint8Array {
  const buf = new Uint8Array(SAVE_SIZE);
  buf[0] = SAVE_MAGIC;
  buf[1] = SAVE_VERSION;
  buf[2] = pet.stage & 0xff;
  buf[3] = pet.hunger & 0xff;
  buf[4] = pet.happiness & 0xff;
  buf[5] = pet.energy & 0xff;
  buf[6] = pet.health & 0xff;
  buf[7] = pet.flags & 0xff;
  writeU32LE(buf, 8, pet.ageTicks >>> 0);
  writeU32LE(buf, 12, pet.careScore >>> 0);
  buf[16] = crc8(buf, 16);
  return buf;
}

export function deserialize(buf: Uint8Array): PetState | null {
  if (buf.length < SAVE_SIZE) return null;
  if (buf[0] !== SAVE_MAGIC || buf[1] !== SAVE_VERSION) return null;
  if (crc8(buf, 16) !== buf[16]) return null;

  const pet = createNewPet();
  pet.stage = buf[2] as Stage;
  pet.hunger = buf[3];
  pet.happiness = buf[4];
  pet.energy = buf[5];
  pet.health = buf[6];
  pet.flags = buf[7];
  pet.ageTicks = readU32LE(buf, 8);
  pet.careScore = readU32LE(buf, 12);
  pet.bootTicks = 0;
  pet.screen = pet.health <= 0 ? Screen.Dead : pet.flags & FLAG_SLEEPING ? Screen.Sleeping : Screen.Home;
  pet.displayOn = true;
  return pet;
}

export function toBase64(buf: Uint8Array): string {
  let s = "";
  for (let i = 0; i < buf.length; i++) s += String.fromCharCode(buf[i]);
  return btoa(s);
}

export function fromBase64(b64: string): Uint8Array | null {
  try {
    const s = atob(b64);
    const buf = new Uint8Array(s.length);
    for (let i = 0; i < s.length; i++) buf[i] = s.charCodeAt(i);
    return buf;
  } catch {
    return null;
  }
}

File: firmware/src/save.c
#include "save.h"
#include <avr/eeprom.h>

uint8_t save_crc8(const uint8_t *data, uint8_t len) {
  uint8_t crc = 0;
  for (uint8_t i = 0; i < len; i++) {
    crc ^= data[i];
    for (uint8_t b = 0; b < 8; b++) {
      crc = (crc & 0x80) ? (uint8_t)((crc << 1) ^ 0x07) : (uint8_t)(crc << 1);
    }
  }
  return crc;
}

void save_write(const Pet *p) {
  uint8_t buf[SAVE_SIZE];
  buf[0] = SAVE_MAGIC;
  buf[1] = SAVE_VERSION;
  buf[2] = p->stage;
  buf[3] = p->hunger;
  buf[4] = p->happiness;
  buf[5] = p->energy;
  buf[6] = p->health;
  buf[7] = p->flags;
  buf[8] = (uint8_t)(p->age_ticks);
  buf[9] = (uint8_t)(p->age_ticks >> 8);
  buf[10] = (uint8_t)(p->age_ticks >> 16);
  buf[11] = (uint8_t)(p->age_ticks >> 24);
  buf[12] = (uint8_t)(p->care);
  buf[13] = (uint8_t)(p->care >> 8);
  buf[14] = (uint8_t)(p->care >> 16);
  buf[15] = (uint8_t)(p->care >> 24);
  buf[16] = save_crc8(buf, 16);
  eeprom_update_block(buf, (void *)0, SAVE_SIZE);
}

uint8_t save_read(Pet *p) {
  uint8_t buf[SAVE_SIZE];
  eeprom_read_block(buf, (const void *)0, SAVE_SIZE);
  if (buf[0] != SAVE_MAGIC || buf[1] != SAVE_VERSION) return 0;
  if (save_crc8(buf, 16) != buf[16]) return 0;
  pet_reset(p);
  p->stage = buf[2];
  p->hunger = buf[3];
  p->happiness = buf[4];
  p->energy = buf[5];
  p->health = buf[6];
  p->flags = buf[7];
  p->age_ticks = (uint32_t)buf[8] | ((uint32_t)buf[9] << 8) |
                 ((uint32_t)buf[10] << 16) | ((uint32_t)buf[11] << 24);
  p->care = (uint32_t)buf[12] | ((uint32_t)buf[13] << 8) |
            ((uint32_t)buf[14] << 16) | ((uint32_t)buf[15] << 24);
  p->screen = (p->health == 0) ? SCR_DEAD : SCR_HOME;
  return 1;
}
